File: docs/DSA/src/py/balanced_binary_tree.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def _helper(self, node):
        if not node:
            return True, 0
        lb, lh = self._helper(node.left)
        rb, rh = self._helper(node.right)
        b = abs(lh-rh) <= 1 and lb and rb
        h = 1+max(lh, rh)
        return b, h

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        balanced, _ = self._helper(root)
        return balanced

File: docs/DSA/src/go/balanced_binary_tree.go
package main

import "math"

func isBalanced(root *TreeNode) bool {
	var helper func(node *TreeNode) (bool, int)
	helper = func(node *TreeNode) (bool, int) {
		if node == nil {
			return true, 0
		}

		lb, lh := helper(node.Left)
		rb, rh := helper(node.Right)
		b := math.Abs(float64(lh)-float64(rh)) <= 1 && lb && rb
		h := 1 + max(lh, rh)
		return b, h
	}

	balanced, _ := helper(root)
	return balanced
}
